drivers/dma/dma_silabs_siwx91x_gpdma.c

```diff
@@ -0,0 +1,540 @@
+ */
+
+#include <zephyr/irq.h>
+#include <zephyr/sys/mem_blocks.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/logging/log.h>
+#include "rsi_gpdma.h"
+#include "rsi_rom_gpdma.h"
+
+#define DT_DRV_COMPAT                silabs_gpdma
+#define GPDMA_DESC_MAX_TRANSFER_SIZE 4096
+#define GPDMA_MAX_CHANNEL_FIFO_SIZE  64
+#define GPDMA_MAX_PRIORITY           3
+#define GPDMA_MIN_PRIORITY           0
+
+LOG_MODULE_REGISTER(si91x_gpdma, CONFIG_DMA_LOG_LEVEL);
+
+enum gpdma_xfer_dir {
+	SIWX91X_TRANSFER_MEM_TO_MEM = 0,
+	SIWX91X_TRANSFER_MEM_TO_PER = 1,
+	SIWX91X_TRANSFER_PER_TO_MEM = 2,
+	SIWX91X_TRANSFER_DIR_INVALID = 4,
+};
+
+enum gpdma_data_width {
+	SIWX91X_DATA_WIDTH_8 = 0,
+	SIWX91X_DATA_WIDTH_16 = 1,
+	SIWX91X_DATA_WIDTH_32 = 2,
+	SIWX91X_DATA_WIDTH_INVALID = 3,
+};
+
+struct siwx91x_gpdma_channel_info {
+	dma_callback_t cb;
+	void *cb_data;
+	RSI_GPDMA_DESC_T *desc;
+	enum gpdma_xfer_dir xfer_direction;
+};
+
+struct siwx91x_gpdma_config {
+	GPDMA_G_Type *reg;         /* GPDMA base address */
+	GPDMA_C_Type *channel_reg; /* GPDMA channel registers address */
+	const struct device *clock_dev;
+	clock_control_subsys_t clock_subsys;
+	void (*irq_configure)(void);
+};
+
+struct siwx19x_gpdma_data {
+	struct dma_context dma_ctx;
+	GPDMA_DATACONTEXT_T hal_ctx;
+	struct sys_mem_blocks *desc_pool;
+	struct k_spinlock desc_pool_lock;
+	struct siwx91x_gpdma_channel_info *chan_info;
+	uint8_t reload_compatible;
+};
+
+static bool siwx91x_gpdma_is_priority_valid(uint32_t channel_priority)
+{
+	return (channel_priority >= GPDMA_MIN_PRIORITY && channel_priority <= GPDMA_MAX_PRIORITY);
+}
+
+static enum gpdma_xfer_dir siwx91x_gpdma_xfer_dir(uint32_t dir)
+{
+	switch (dir) {
+	case MEMORY_TO_MEMORY:
+		return SIWX91X_TRANSFER_MEM_TO_MEM;
+	case MEMORY_TO_PERIPHERAL:
+		return SIWX91X_TRANSFER_MEM_TO_PER;
+	case PERIPHERAL_TO_MEMORY:
+		return SIWX91X_TRANSFER_PER_TO_MEM;
+	default:
+		return SIWX91X_TRANSFER_DIR_INVALID;
+	}
+}
+
+static enum gpdma_data_width siwx91x_gpdma_data_size(uint32_t data_size)
+{
+	switch (data_size) {
+	case 1:
+		return SIWX91X_DATA_WIDTH_8;
+	case 2:
+		return SIWX91X_DATA_WIDTH_16;
+	case 4:
+		return SIWX91X_DATA_WIDTH_32;
+	default:
+		return SIWX91X_DATA_WIDTH_INVALID;
+	}
+}
+
+static int siwx91x_gpdma_ahb_blen(uint8_t blen)
+{
+	if (blen == 1) {
+		return AHBBURST_SIZE_1;
+	} else if (blen <= 4) {
+		return AHBBURST_SIZE_4;
+	} else if (blen <= 8) {
+		return AHBBURST_SIZE_8;
+	} else if (blen <= 16) {
+		return AHBBURST_SIZE_16;
+	} else if (blen <= 20) {
+		return AHBBURST_SIZE_20;
+	} else if (blen <= 24) {
+		return AHBBURST_SIZE_24;
+	} else if (blen <= 28) {
+		return AHBBURST_SIZE_28;
+	} else if (blen <= 32) {
+		return AHBBURST_SIZE_32;
+	} else {
+		return -EINVAL;
+	}
+}
+
+static void siwx91x_gpdma_free_desc(sys_mem_blocks_t *mem_block, void *block)
+{
+	RSI_GPDMA_DESC_T *cur_desc = (RSI_GPDMA_DESC_T *)block;
```

Comment:
`siwx91x_gpdma_free_desc()` is always called with a `data->chan_info[channel].desc)` as parameter. You can change the type of `block` in a `RSI_GPDMA_DESC_T *`, remove all the useless casts and rename `block` in `cur_desc`.